### PR TITLE
adding titanic example

### PR DESCRIPTION
**Pull request by Arvind-Ideas2IT**
Signed-off-by: Arvind Gautam <arvindkumarsingh.gautam@ideas2it.com>

## What changes are proposed in this pull request?
Adding a simple example with explanations API for Captum interpretability of the model on the titanic data set. 

## How is this patch tested?
Tested on the local machine. and all the unit test got passed

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow TorchServe Deployment Plugin users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s) does this PR affect?
Components
- [ ] `area/deploy`: Main deployment plugin logic
- [ ] `area/build`: Build and test infrastructure for MLflow TorchServe Deployment Plugin
- [ ] `area/docs`: MLflow TorchServe Deployment Plugin documentation pages
- [x] `area/examples`: Example code

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: examples/Titanic/conda.yaml [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+- conda-forge
+dependencies:
+- python=3.8.2
+- pytorch
+- pip
+- pip:
+  - mlflow
+  - pandas
+  - captum
+  - boto3
```
Comment (shrinath-suresh) [inline examples/Titanic/conda.yaml:13]: Following package is not used in the example
```suggestion
```

File: examples/Titanic/config.properties [hunk 2]
```diff
@@ -0,0 +1,4 @@
+number_of_netty_threads=4
```
Comment (shrinath-suresh) [inline examples/Titanic/config.properties:1]: Please remove this file.

File: examples/Titanic/test_data/titanic_survived.csv [hunk 3]
```diff
@@ -0,0 +1,1310 @@
+age,sibsp,parch,fare,female,male,embark_C,embark_Q,embark_S,class_1,class_2,class_3
+29.0,0,0,211.3375,1,0,0,0,1,1,0,0
+0.92,1,2,151.55,0,1,0,0,1,1,0,0
```
Comment (shrinath-suresh) [inline examples/Titanic/test_data/titanic_survived.csv:3]: Lets keep first 5 or 10 rows . Its just a sample input to run the file. Please add the link to dataset in the README.md file if in case user wants to access it.

File: examples/Titanic/titanic_captum_interpret.py [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+from scipy import stats
+import pandas as pd
+import mlflow
+from prettytable import PrettyTable
+from sklearn.model_selection import train_test_split
+import os
+from argparse import ArgumentParser
+import torch.nn as nn
+
+mlflow.start_run(run_name="Titanic_Captum_mlflow")
```
Comment (shrinath-suresh) [inline examples/Titanic/titanic_captum_interpret.py:20]:
```suggestion
with mlflow.start_run(run_name="Titanic_Captum_mlflow"):
```

File: examples/Titanic/titanic_captum_interpret.py [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+        ],
+        axis=1,
+    )
+    return titanic_data
+
+
+torch.manual_seed(1)  # Set seed for reproducibility.
+
+
+class TitanicSimpleNNModel(nn.Module):
```
Comment (shrinath-suresh) [inline examples/Titanic/titanic_captum_interpret.py:77]:
```suggestion

```
This class already available as part of model file. Please remove the duplicate classes

File: examples/Titanic/titanic_captum_interpret.py [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+        data, labels, test_size=0.3, random_state=RANDOM_SEED, stratify=labels
+    )
+    return (train_features, train_labels, test_features, test_labels, feature_names)
+
+
+def count_model_parameters(model):
+    table = PrettyTable(["Modules", "Parameters"])
+    total_params = 0
+    for name, parameter in model.named_parameters():
+
```
Comment (shrinath-suresh) [inline examples/Titanic/titanic_captum_interpret.py:112]:
```suggestion

```
nit

File: examples/Titanic/titanic_captum_interpret.py [hunk 7]
```diff
@@ -0,0 +1,375 @@
...
+
+def visualize_importances(
+    feature_names,
+    importances,
+    title="Average Feature Importances",
+    plot=True,
+    axis_title="Features",
+):
+    print(title)
```
Comment (shrinath-suresh) [inline examples/Titanic/titanic_captum_interpret.py:129]:
```suggestion
```

File: examples/Titanic/titanic_handler.py [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+        feature_imp_dict = {}
+        for i in range(len(self.feature_names)):
+            feature_imp_dict[str(self.feature_names[i])] = importances[i]
+        x_pos = np.arange(len(self.feature_names))
+        fig, ax = plt.subplots(figsize=(12, 6))
+        ax.bar(x_pos, importances, align="center")
+        ax.set(title=title, xlabel=axis_title)
+        ax.set_xticks(x_pos)
+        ax.set_xticklabels(self.feature_names, rotation="vertical")
+        print("Input file path: ", self.input_file_path)
```
Comment (shrinath-suresh) [inline examples/Titanic/titanic_handler.py:143]:
```suggestion

```

File: examples/Titanic/README.md [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+curl http://127.0.0.1:8080/explanations/titanic --data '{"input_file_path" : "/home/ubuntu/titanic/data/titanic_survived.csv"}'
+
+
+### Captum Explanations
+
+The explain is called with the following request api http://127.0.0.1:8080/explanations/titanic
+
+Captum/Explain doesn't support batching.
+
+#### The handler changes:
```
Comment (shrinath-suresh) [inline examples/Titanic/README.md:67]: This section can be removed

File: examples/Titanic/README.md [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    ```
+
+For captum Explanations on the Torchserve side, use the below curl request:
+```bash
+curl http://127.0.0.1:8080/explanations/titanic --data '{"input_file_path" : "/home/ubuntu/titanic/data/titanic_survived.csv"}'
+
+
+### Captum Explanations
+
+The explain is called with the following request api http://127.0.0.1:8080/explanations/titanic
```
Comment (shrinath-suresh) [inline examples/Titanic/README.md:63]: Can you please add whats the expected output

File: examples/Titanic/README.md [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+    torch-model-archiver --model-name titanic --version 1.0 --model-file titanic.py --serialized-file titanic_state_dict.pt  --handler  titanic_handler.py --extra-file index_to_name.json
+    ```
+   
+ * Step - 5: Register the model on TorchServe using the above model archive file and run inference
+   
+    ```bash
+    mkdir model_store
+    mv titanic.mar model_store/
+     torchserve --start --model-store model_store/ --ts-config config.properties
+    curl http://127.0.0.1:8080/predictions/titanic --data '{"input_file_path" : "/home/ubuntu/titanic/test_data/titanic_survived.csv"}'
```
Comment (shrinath-suresh) [inline examples/Titanic/README.md:53]: Can you please add what is the expected output of prediction

**Review by shrinath-suresh — CHANGES_REQUESTED** [review]

File: examples/Titanic/README.md (hunk 11 above)
Comment (Arvind-Ideas2IT) [reply examples/Titanic/README.md:53]: @shrinath-suresh  I have addressed your review comments.

File: examples/Titanic/titanic_handler.py [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+        """Predict the class (survived or not survived) of the received input dataframe using the
+        serialized model.
+
+        Args:
+            input_batch (list): List of Text Tensors from the pre-process function is passed here
+
+        Returns:
+            list : It returns a list of the predicted value for the input text
+        """
+        self.out_probs = self.model(data[0]).detach().numpy()
```
Comment (HamidShojanazeri) [inline examples/Titanic/titanic_handler.py:98]: @Arvind-Ideas2IT setting device to GPU, this requires the data to be moved to right device. 
In case to prevent from converting it to cpu for np.argmax,  "self.out_probs.argmax(1).item()", could be used as well.

File: examples/Titanic/README.md [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+```
+mlflow run . --no-conda
+```
+
+# Above commands will train the titanic model for further use.
+
+
+# Serve a custom model on TorchServe
+
+Run the commands given in following steps from the example dir. For example, if you cloned the repository into /home/my_path/serve, run the steps from /home/my_path/CaptumExample
```
Comment (HamidShojanazeri) [inline examples/Titanic/README.md:37]: @Arvind-Ideas2IT Thanks for the PR,  would you please clarify on this, now the example is residing on "/home/ubuntu/mlflow-torchserve/examples/Titanic" should the commands be run from "/home/ubuntu/mlflow-torchserve/examples"?

**Review by HamidShojanazeri — COMMENTED** [review]
Please refer to inline comments.

File: examples/Titanic/README.md (hunk 11 above)
Comment (Arvind-Ideas2IT) [reply examples/Titanic/README.md:53]: @HamidShojanazeri Thanks for catching this out. You did not miss any step :-). I have updated the code block to use JSON as input, and you don't need to add  "service_envelope=body" in config. properties. My apologies, I pushed the update code in some other PR previously

File: examples/Titanic/README.md [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+
+ * Step - 1: Create a new model architecture file which contains model class extended from torch.nn.modules. In this example we have created [titanic model file](titanic.py).
+ * Step - 2: Write a custom handler to run the inference on your model. In this example, we have added a [custom_handler](titanic_handler.py) which runs the inference on the input record using the above model and make prediction.
+ * Step - 3: Create a torch model archive using the torch-model-archiver utility to archive the above files.
+ 
+    ```bash
+    torch-model-archiver --model-name titanic --version 1.0 --model-file titanic.py --serialized-file titanic_state_dict.pt  --handler  titanic_handler.py --extra-file index_to_name.json
```
[the remark this answers is not in the document]
Comment (Arvind-Ideas2IT) [reply examples/Titanic/README.md:44]: @HamidShojanazeri , Thanks for pointing it out. I have updated the README

File: examples/Titanic/titanic_handler.py (hunk 12 above)
Comment (Arvind-Ideas2IT) [reply examples/Titanic/titanic_handler.py:98]: @HamidShojanazeri  Thanks for the suggestion. I will add this

File: examples/Titanic/README.md (hunk 13 above)
Comment (Arvind-Ideas2IT) [reply examples/Titanic/README.md]: @HamidShojanazeri  Sorry, I have updated the README. The command should run from ./examples/Titanic dir.

**Comment by Arvind-Ideas2IT** [comment]
@HamidShojanazeri I have addressed the comments and pushed the updated PR, Please review it whenever you get time.

File: examples/Titanic/titanic_handler.py [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+    def inference(self, data):
+        """Predict the class (survived or not survived) of the received input json file using the
+        serialized model.
+
+        Args:
+            input_batch (list): List of Tensors from the pre-process function is passed here
+
+        Returns:
+            list : It returns a list of the predicted value for the input test record
+        """
```
Comment (HamidShojanazeri) [inline examples/Titanic/titanic_handler.py:96]: @Arvind-Ideas2IT  Please move data to the device, data = data.to(self.device)

File: examples/Titanic/titanic_handler.py [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+        self.out_probs = self.model(data).detach().numpy()
```
Comment (HamidShojanazeri) [inline examples/Titanic/titanic_handler.py:97]: @Arvind-Ideas2IT please remove .detach().numpy() as it needs the input to be on cpu.

File: examples/Titanic/titanic_handler.py [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+    ):
+        """This function calls the integrated gradient to the feature importance
+
+        Args:
+            data(tensor):
+            target (int): The Target can be set to any acceptable label under the user's discretion.
+
+        Returns:
+            (list): Returns a dict of feature names and their importances
+        """
```
Comment (HamidShojanazeri) [inline examples/Titanic/titanic_handler.py:125]: @Arvind-Ideas2IT Please move the input to device,  input_tensor.to(self.device)

File: examples/Titanic/titanic_handler.py [hunk 18]
```diff
@@ -0,0 +1,158 @@
...
+        ig = IntegratedGradients(self.model)
+        input_tensor.requires_grad_()
+        attr, self.delta = ig.attribute(input_tensor, target=1, return_convergence_delta=True)
+        attr = attr.detach().numpy()
```
Comment (HamidShojanazeri) [inline examples/Titanic/titanic_handler.py:129]: @Arvind-Ideas2IT as input has been moved to device, need to move it to cpu for converting to numpy, attr = attr.cpu().detach().numpy()

**Review by HamidShojanazeri — COMMENTED** [review]
@Arvind-Ideas2IT Thanks for addressing previous comments, please refer to in-line comments.

File: examples/Titanic/titanic_handler.py (hunk 15 above)
Comment (Arvind-Ideas2IT) [reply examples/Titanic/titanic_handler.py:96]: @HamidShojanazeri Sure, I will do that

File: examples/Titanic/titanic_handler.py (hunk 18 above)
Comment (Arvind-Ideas2IT) [reply examples/Titanic/titanic_handler.py:129]: @HamidShojanazeri Got it,Thanks. I will update these changes.

**Comment by Arvind-Ideas2IT** [comment]
@HamidShojanazeri  I have addressed your latest comments,  all the checks are also passed. Please review it.

**Comment by HamidShojanazeri** [comment]
@Arvind-Ideas2IT Thanks LGTM, just please add a note in the README to modify/pass the path to the csv file in the input.json.

**Comment by Arvind-Ideas2IT** [comment]
> @Arvind-Ideas2IT Thanks LGTM, just please add a note in the README to modify/pass the path to the csv file in the input.json.
 
@HamidShojanazeri  Sure, I will do that. Thanks.

**Comment by Arvind-Ideas2IT** [comment]
@HamidShojanazeri  I have updated the READ.md as you suggested, removed the absolute path of the CSV in the input.json

File: examples/Titanic/config.properties (hunk 2 above)
Comment (Arvind-Ideas2IT) [reply examples/Titanic/config.properties]: @shrinath-suresh  Done that

**Review by shrinath-suresh — APPROVED** [review]
LGTM

**Review by chauhang — APPROVED** [review]
